addons/purchase_stock/models/purchase.py

```diff
@@ -272,7 +272,13 @@ def _get_orders_to_remind(self):
 class PurchaseOrderLine(models.Model):
     _inherit = 'purchase.order.line'
 
-    qty_received_method = fields.Selection(selection_add=[('stock_moves', 'Stock Moves')])
+    def _ondelete_stock_moves(self):
+        for record in self:
+            record.qty_received_manual = record.qty_received
+            record.qty_received_method = 'manual'
```

Comment:
Using the `ondelete` mechanism is quite nice (it was designed for that purpose, after all.)
However, I am wondering about performance: on a large database, this may never end.
What about doing it in plain SQL instead?
The only potential issue would be dependent computed fields that would not be recomputed...

Reply:
Yes, I thought about this, but you're also mentioning the exact reason why I didn't. I thought about it some more though, and I'm wondering about these options:
1) What if we updated `qty_received_manual` through sql? And then do a single `write` to set the `qty_received_method`? Because the only compute depending on `qty_received_method` also depends on `qty_received_manual` (it is the field `qty_received`) we could probably use the following strategy to avoid a linear amount of queries: simply check if the computed value is already correct, if it is don't write it.
    ```python
    @api.depends('qty_received_method', 'qty_received_manual')
    def _compute_qty_received(self):
        lines_manual = self.filtered(lambda line: (
            line.qty_received_method == 'manual'
            and line.qty_received != line.qty_received_manual
        ))
        # This is a zero-length for loop when triggering the sql based ondelete.
        for line in lines_manual:
            line.qty_received = line.qty_received_manual or 0.0

        self.filtered(lambda line: line.qty_received_method != 'manual').write({
            'qty_received': 0.0,
        })
    ```
2) Can we trigger computes manually? If yes, combining this with the strategy above ensures correctness in all cases (including non-odoo custo computes depending on these fields).